grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -1358,4 +1361,37 @@ class ProductService {
 
         return productInstance
     }
+
+    void sendProductAvailabilityMessages(String locationId, List productIds){
+        Location location = Location.get(locationId)
+        List<Product> products = Product.findAllByIdInList(productIds)
+        log.info "location:${location} from locationId:${locationId}, products:${products} from ids:${productIds}"
+        Map<Product, Integer> quantityMap = productAvailabilityService.getQuantityOnHandByProduct(location, products)
+        Map<Product, Integer> quantityAvailableMap = productAvailabilityService.getQuantityAvailableToPromiseByProduct(location, products)
+        List availableItems = productAvailabilityService.getAvailableItems(location, products)
+        sendProductAvailabilityMessages(location, products, quantityMap, quantityAvailableMap, availableItems)
+    }
+
+    void sendProductAvailabilityMessages(Location location, List<Product> products, Map<Product, Integer> quantityMap,
+                                         Map<Product, Integer> quantityAvailableMap, List availableItems){
+        String TOPIC_ARN = grailsApplication.config.awssdk.sns.productAvailability
+        Attribute externalIdAttribute = findExternalProductIdAttribute()
+        JSONArray jsonArray = new JSONArray()
+        products?.each {Product product ->
+            JSONObject jsonObject = new JSONObject()
+            String externalId = product.attributes.find{it.attribute == externalIdAttribute}.value
+            List productAvailableItem = availableItems.findAll {it.product == product}
+            jsonObject.put("productId", externalId)
+            jsonObject.put("productCode", product?.productCode)
+            jsonObject.put("locationNumber", location?.locationNumber)
+            jsonObject.put("quantityOnHand", quantityMap[product])
+            jsonObject.put("quantityAvailable", quantityAvailableMap[product])
+            jsonObject.put("earliestExpirationDate", productAvailableItem?.min {it?.inventoryItem?.expirationDate})
```

Comment:
We should be able to get quantity on hand, quantity available, and min expiration date in a single query. We can leave this for now, but I'll try to refactor when I get a chance.